openmdao/components/tests/test_exec_comp.py

```diff
@@ -1196,6 +1218,320 @@ def test_list_outputs_resids_tol(self):
+            self.assertEquals(cm.exception.args[0],
+                              "'comp' <class ExecComp>: expression contains functions ['unsafe'] that are not complex safe. To fix this, call declare_partials('*', ['z'], method='fd') on this component prior to setup.")
+
+    def test_register_check_partials_safe(self):
+        with _temporary_expr_dict():
+            size = 10
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)', shape=size))
+            p.setup()
+            p['comp.x'] = 3.
+            p.run_model()
+            assert_near_equal(p['comp.area_square'], np.ones(size) * 9., 1e-6)
+
+            data = p.check_partials(out_stream=None)
+            self.assertEqual(list(data), [])
+
+    def test_register_simple_arr_manual_partials_cs(self):
+        with _temporary_expr_dict():
+            size = 10
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            comp = p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)', shape=size))
+            comp.declare_partials('area_square', 'x', method='cs')
+            p.setup()
+            p['comp.x'] = 3.
+            p.run_model()
+            assert_near_equal(p['comp.area_square'], np.ones(size) * 9., 1e-6)
+            J = p.compute_totals(of=['comp.area_square'], wrt=['comp.x'])
+            assert_near_equal(J['comp.area_square', 'comp.x'], np.eye(size) * 6., 1e-6)
+
+    def test_register_simple_arr_manual_partials_fd(self):
+        with _temporary_expr_dict():
+            size = 10
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=False)
+            p = om.Problem()
+            comp = p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)', shape=size))
+            comp.declare_partials('area_square', 'x', method='fd')
+            p.setup()
+            p['comp.x'] = 3.
+            p.run_model()
+            assert_near_equal(p['comp.area_square'], np.ones(size) * 9., 1e-6)
+            J = p.compute_totals(of=['comp.area_square'], wrt=['comp.x'])
+            assert_near_equal(J['comp.area_square', 'comp.x'], np.eye(size) * 6., 1e-6)
+
+    def test_register_simple_arr_diag(self):
+        with _temporary_expr_dict():
+            size = 10
+            om.ExecComp.register('area', lambda x: x**2, complex_safe=True)
+            p = om.Problem()
+            p.model.add_subsystem('comp', om.ExecComp('area_square = area(x)', shape=size, has_diag_partials=True))
+            p.setup()
+            p['comp.x'] = 3.
+            p.run_model()
+            assert_near_equal(p['comp.area_square'], np.ones(size) * 9., 1e-6)
+            J = p.compute_totals(of=['comp.area_square'], wrt=['comp.x'])
+            assert_near_equal(J['comp.area_square', 'comp.x'], np.eye(size) * 6., 1e-6)
```

Comment:
I'm not sure what this test does. It's not showing that has_diag_partials is doing anything different (and you'd only be able to tell by looking at the size of rows/cols/val for that subjac anyway).  I guess maybe it just proves that the registered function doesn't break the code that runs with has_diag_partials.

Reply:
fixed